File: samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/handlers/EnsureSupportUserActorHandler.java
package systems.zlink.samples.supportchat.server.support.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.actors.ActorRef;
import systems.zlink.framework.actors.ActorRefSnapshot;
import systems.zlink.framework.actors.ZLinkActorCreateResult;
import systems.zlink.framework.actors.ZLinkActorManager;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.supportchat.server.configuration.SampleNames;
import systems.zlink.samples.supportchat.shared.contracts.Messages;

import java.util.concurrent.CompletionStage;

@ZLinkHandlerGroup(SampleNames.SupportChannel)
public final class EnsureSupportUserActorHandler
        implements ZLinkRequestHandler<
                Messages.EnsureSupportUserActorReq, Messages.EnsureSupportUserActorRes> {
    private final ZLinkActorManager actors;

    public EnsureSupportUserActorHandler(ZLinkActorManager actors) {
        this.actors = actors;
    }

    @Override
    public CompletionStage<Messages.EnsureSupportUserActorRes> handle(
            Messages.EnsureSupportUserActorReq request, ZLinkMessageContext context) {
        return actors.getOrCreate(request.actorId(), SampleNames.SupportActorType)
                .request(request)
                .submit()
                .thenApply(
                        result ->
                                new Messages.EnsureSupportUserActorRes(
                                        ActorRefSnapshot.from(actorRef(result))));
    }

    private static ActorRef actorRef(ZLinkActorCreateResult result) {
        if (result instanceof ZLinkActorCreateResult.Created created) {
            return created.actor();
        }
        if (result instanceof ZLinkActorCreateResult.Existing existing) {
            return existing.actor();
        }
        throw new IllegalStateException("Support actor creation was rejected");
    }
}

File: samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/infrastructure/zlink/handlers/EnsureSupportUserActorHandler.kt
package systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.handlers

import kotlinx.coroutines.future.await
import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.actors.ActorRef
import systems.zlink.framework.actors.ActorRefSnapshot
import systems.zlink.framework.actors.ZLinkActorCreateResult
import systems.zlink.framework.actors.ZLinkActorManager
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.framework.kotlin.kotlin
import systems.zlink.samples.kotlin.supportchat.server.configuration.SampleNames
import systems.zlink.samples.kotlin.supportchat.shared.contracts.EnsureSupportUserActorReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.EnsureSupportUserActorRes

@ZLinkHandlerGroup(SampleNames.SupportChannel)
class EnsureSupportUserActorHandler(private val actors: ZLinkActorManager) :
    ZLinkSuspendingRequestHandler<EnsureSupportUserActorReq, EnsureSupportUserActorRes> {
    override suspend fun handle(
        request: EnsureSupportUserActorReq,
        context: ZLinkMessageContext,
    ): EnsureSupportUserActorRes {
        val existing = actors.find(request.actorId).await().orElse(null)
        if (existing != null) {
            return existing.toResponse()
        }
        val actor =
            actors
                .kotlin()
                .getOrCreate(request.actorId, SampleNames.SupportActorType)
                .request(request)
                .await()
        return actor.requireActor().toResponse()
    }
}

private fun ActorRef.toResponse(): EnsureSupportUserActorRes =
    EnsureSupportUserActorRes(actor = ActorRefSnapshot.from(this))

private fun ZLinkActorCreateResult.requireActor(): ActorRef =
    when (this) {
        is ZLinkActorCreateResult.Created -> actor
        is ZLinkActorCreateResult.Existing -> actor
        is ZLinkActorCreateResult.Rejected -> error("Support actor creation was rejected")
    }
